File: java-algorithm/src/main/java/com/leetcode/Problem600.java
package com.leetcode;

/**
 * 600. Non-negative Integers without Consecutive Ones
 * https://leetcode.com/problems/non-negative-integers-without-consecutive-ones/
 */
public class Problem600 {
    public int findIntegers(int n) {
        int len = getLen(n);
        return dfs(n, len, 0, 0);
    }

    private int dfs(int n, int len, int pos, int bits) {
        if (pos == len)
            return 1;
        int ret = 0;
        int pre_bit = bits & 1;
        bits = bits << 1;
        if (bits <= n)
            ret += dfs(n, len, pos + 1, bits);
        if (pre_bit == 0 && bits + 1 <= n)
            ret += dfs(n, len, pos + 1, bits + 1);
        return ret;
    }

    private int getLen(int n) {
        for (int i = 31; i >= 0; i--) {
            if (((n >> i) & 1) == 1)
                return i;
        }
        return 0;
    }
}

File: python-algo/leetcode/problem_600.py
"""600. Non-negative Integers without Consecutive Ones
https://leetcode.com/problems/non-negative-integers-without-consecutive-ones/
"""
import math
from functools import lru_cache


class Solution:
    def find_integers(self, n: int) -> int:
        @lru_cache(None)
        def dfs(pos: int, bits: int) -> int:
            if pos == m:
                return 1
            last_bit = bits & 1
            ret = 0
            bits = bits << 1
            if bits <= n:  # cur pos pick 0
                ret += dfs(pos + 1, bits)
            if last_bit == 0 and bits + 1 <= n:  # cur pos pick 1
                ret += dfs(pos + 1, bits + 1)
            return ret

        dp = [[0, 0] for _ in range(31)]
        dp[1][0] = 1
        dp[1][1] = 2
        for i in range(1, 30):
            dp[i + 1][0] = dp[i][1]
            dp[i + 1][1] = dp[i][0] + dp[i][1]
        ans = 0
        pre = 0
        for i in range(30, -1, -1):
            bit = (n >> i) & 1
            if bit == 1:
                ans += dp[i + 1][0]
                if pre == 1:
                    break
            pre = bit
            if i == 0:
                ans += 1
        return ans
